akka-remote/src/main/scala/akka/remote/serialization/MiscMessageSerializer.scala

```diff
@@ -0,0 +1,117 @@
+ */
+package akka.remote.serialization
+
+import akka.actor._
+import akka.protobuf.ByteString
+import akka.remote.ContainerFormats
+import akka.serialization.{ BaseSerializer, SerializationExtension, SerializerWithStringManifest }
+
+class MiscMessageSerializer(val system: ExtendedActorSystem) extends SerializerWithStringManifest with BaseSerializer {
+
+  @deprecated("Use constructor with ExtendedActorSystem", "2.4")
+  def this() = this(null)
+
+  private lazy val serialization = SerializationExtension(system)
+
+  // TODO remove this when deprecated this() is removed
+  override val identifier: Int =
+    if (system eq null) 9
+    else identifierFromConfig
+
+  def toBinary(obj: AnyRef): Array[Byte] = obj match {
+    case identify: Identify      ⇒ serializeIdentify(identify)
+    case identity: ActorIdentity ⇒ serializeActorIdentity(identity)
+    case _                       ⇒ throw new IllegalArgumentException(s"Cannot serialize object of type [${obj.getClass.getName}]")
+  }
+
+  private def serializeIdentify(identify: Identify): Array[Byte] =
+    ContainerFormats.Identify.newBuilder()
+      .setMessageId(payloadBuilder(identify.messageId))
+      .build()
+      .toByteArray
+
+  private def serializeActorIdentity(actorIdentity: ActorIdentity): Array[Byte] = {
+    val builder =
+      ContainerFormats.ActorIdentity.newBuilder()
+        .setCorrelationId(payloadBuilder(actorIdentity.correlationId))
+
+    actorIdentity.ref.foreach { actorRef ⇒
+      builder.setRef(actorRefBuilder(actorRef))
+    }
+
+    builder
+      .build()
+      .toByteArray
+  }
+
+  private def actorRefBuilder(actorRef: ActorRef): ContainerFormats.ActorRef.Builder =
+    ContainerFormats.ActorRef.newBuilder()
+      .setPath(actorRef.path.toString)
+
+  private def payloadBuilder(input: Any): ContainerFormats.Payload.Builder = {
+    val payload = input.asInstanceOf[AnyRef]
+    val builder = ContainerFormats.Payload.newBuilder()
+    val serializer = serialization.findSerializerFor(payload)
+
+    builder
+      .setEnclosedMessage(ByteString.copyFrom(serializer.toBinary(payload)))
+      .setSerializerId(serializer.identifier)
+
+    serializer match {
+      case ser2: SerializerWithStringManifest ⇒
+        val manifest = ser2.manifest(payload)
+        if (manifest != "")
+          builder.setMessageManifest(ByteString.copyFromUtf8(manifest))
+      case _ ⇒
+        if (serializer.includeManifest)
+          builder.setMessageManifest(ByteString.copyFromUtf8(payload.getClass.getName))
+    }
+
+    builder
+  }
+
+  override def manifest(o: AnyRef): String =
+    o.getClass.getName
+
+  private val IdentifyClassName = classOf[Identify].getName
+  private val ActorIdentityClassName = classOf[ActorIdentity].getName
+
+  override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef =
+    manifest match {
+      case IdentifyClassName      ⇒ deserializeIdentify(bytes)
+      case ActorIdentityClassName ⇒ deserializeActorIdentity(bytes)
```

Comment:
we should use short manifest strings instead of class names, "A", "B", see for example https://github.com/akka/akka/blob/master/akka-cluster-tools/src/main/scala/akka/cluster/pubsub/protobuf/DistributedPubSubMessageSerializer.scala#L48